File: backend/app/schemas/ai.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel


class AIGenerateRequest(BaseModel):
    novel_id: UUID
    chapter_id: Optional[UUID] = None
    section_id: Optional[UUID] = None
    prompt: str
    context_type: str  # character, world, plot, content, chapter, outline
    max_tokens: Optional[int] = 2000
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"  # openai, anthropic, custom
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AIGenerateResponse(BaseModel):
    content: str
    tokens_used: int
    model: str


class AICharacterGenerateRequest(BaseModel):
    novel_id: UUID
    character_role: str  # protagonist, antagonist, supporting
    character_traits: Optional[str] = None
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AIPlotGenerateRequest(BaseModel):
    novel_id: UUID
    plot_type: str  # main, subplot, twist
    plot_length: Optional[str] = "medium"  # short, medium, long
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AIChapterOutlineRequest(BaseModel):
    novel_id: UUID
    chapter_number: int
    chapter_theme: Optional[str] = None
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AIContentExpandRequest(BaseModel):
    novel_id: UUID
    # Accept numeric chapter ID (primary key) or string; frontend passes route param string
    chapter_id: Optional[int | str] = None
    content_snippet: str
    expansion_style: Optional[str] = "detailed"  # brief, detailed, dramatic
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AIModelConfigBase(BaseModel):
    provider: str  # openai, anthropic, custom
    base_url: Optional[str] = None
    api_key: str
    model_name: str
    fallback_provider: Optional[str] = None


class AIModelConfigCreate(AIModelConfigBase):
    pass


class AIModelConfigResponse(AIModelConfigBase):
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


class AIWorldGenerateRequest(BaseModel):
    novel_id: UUID
    focus: Optional[str] = None  # era, rules, locations, culture, or overall
    temperature: Optional[float] = None
    provider: Optional[str] = "openai"
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = "gpt-4"


class AITestRequest(BaseModel):
    provider: str  # openai, anthropic, custom
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    model_name: Optional[str] = None


class AITestResponse(BaseModel):
    ok: bool
    provider: str
    message: str

File: frontend/src/types/ai.ts
// Shared AI types to align frontend <-> backend contract

export type Provider = 'openai' | 'anthropic' | 'custom'

export interface AIConfig {
  provider: Provider
  model_name: string
  base_url?: string
  api_key?: string
  temperature?: number
  max_tokens?: number
  fallback_provider?: Provider
}

export type AIContextType = 'character' | 'world' | 'plot' | 'content' | 'chapter' | 'outline'

export interface AIGenerateRequest {
  novel_id: string
  prompt: string
  context_type: AIContextType
  max_tokens?: number
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AIGenerateResponse {
  content: string
  tokens_used: number
  model: string
}

export interface AICharacterGenerateRequest {
  novel_id: string
  character_role: 'protagonist' | 'antagonist' | 'supporting'
  character_traits?: string
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AIPlotGenerateRequest {
  novel_id: string
  plot_type: 'main' | 'subplot' | 'twist'
  plot_length?: 'short' | 'medium' | 'long'
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AIChapterOutlineRequest {
  novel_id: string
  chapter_number: number
  chapter_theme?: string
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AIContentExpandRequest {
  novel_id: string
  // Backend accepts numeric chapter ID primary key; use string | number to match router params
  chapter_id?: string | number
  content_snippet: string
  expansion_style?: 'brief' | 'detailed' | 'dramatic'
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AIWorldGenerateRequest {
  novel_id: string
  focus?: 'era' | 'rules' | 'locations' | 'culture' | 'overall'
  temperature?: number
  provider?: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AITestRequest {
  provider: Provider
  base_url?: string
  api_key?: string
  model_name?: string
}

export interface AITestResponse {
  ok: boolean
  provider: string
  message: string
}

// AI Assistant types
export type AssistantRole = 'conceptualizer' | 'blueplanner' | 'outliner' | 'novelist' | 'extractor' | 'evaluator'

export interface AssistantInfo {
  role: AssistantRole
  name: string
  description: string
}

export interface AssistantRequest {
  role: AssistantRole
  novel_id: string
  user_input: string
  provider?: Provider
  model_name?: string
  base_url?: string
  api_key?: string
  temperature?: number
  max_tokens?: number
}

export interface AssistantResponse {
  role: AssistantRole
  content: string
  tokens_used: number
}

export interface MultipleVersionsResponse {
  role: 'novelist'
  versions: string[]
  count: number
}

